File: src/niconico.ts
import protobuf from 'protobufjs';

export class NiconicoClient {
    private active = false;
    private ws: WebSocket | null = null;
    private onMessage: (msg: any) => void;
    private ndgrRoot: protobuf.Root | null = null;
    private abortController: AbortController | null = null;

    // State
    private currentAt: string | number = 'now';
    private viewUri: string | null = null;

    // Limits
    private maxBackwardSegments = 10;

    constructor(onMessage: (msg: any) => void) {
        this.onMessage = onMessage;
        this.loadProto();
    }

    private async loadProto() {
        try {
            // Use the bundled proto (contains all definitions)
            this.ndgrRoot = await protobuf.load('/ndgr_full.proto');
            console.log('[NicoClient] Proto loaded');
        } catch (e) {
            console.error('[NicoClient] Proto load error', e);
            this.onMessage({ author: 'System', message: 'Failed to load Protocol Definitions.' });
        }
    }

    async join(url: string) {
        if (this.active) return;
        this.active = true;

        try {
            console.log(`[NicoClient] Fetching page: ${url}`);
            const res = await fetch(url);
            const html = await res.text();

            let props;
            const match = html.match(/id="embedded-data".*?data-props="([^"]+)"/);
            if (match) {
                props = JSON.parse(decodeURIComponent(match[1]).replace(/&quot;/g, '"'));
            } else {
                const jsonMatch = html.match(/<script id="embedded-data" type="application\/json">([\s\S]*?)<\/script>/);
                if (jsonMatch) props = JSON.parse(jsonMatch[1]);
            }

            if (!props) throw new Error('Could not find embedded-data');

            // Debug props
            console.log(`[NicoClient] Props Program ID: ${props.program?.nicoliveProgramId}`);

            const wsUrl = props.site?.relive?.webSocketUrl;
            if (!wsUrl) throw new Error('No WebSocket URL found');

            console.log(`[NicoClient] WS URL: ${wsUrl}`);
            this.connectWs(wsUrl);

        } catch (e) {
            console.error('[NicoClient] Join Error', e);
            this.onMessage({ author: 'System', message: `Join Error: ${e}` });
            this.active = false;
        }
    }

    private connectWs(wsUrl: string) {
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
            console.log('[NicoClient] WS Open');
            this.onMessage({ author: 'System', message: `Connected to Niconico WS` });

            // Reference: Send startWatching
            this.ws?.send(JSON.stringify({
                "type": "startWatching",
                "data": {
                    "stream": {
                        "quality": "super_high",
                        "latency": "low",
                        "chasePlay": false
                    },
                    "reconnect": false
                }
            }));
        };

        this.ws.onmessage = (event) => {
            try {
                const json = JSON.parse(event.data.toString());

                if (json.type === 'ping') {
                    this.ws?.send(JSON.stringify({ type: 'pong' }));
                }

                if (json.type === 'messageServer' && json.data?.viewUri) {
                    console.log('[NicoClient] Received MessageServer ViewURI');
                    this.viewUri = json.data.viewUri;
                    this.startEntryLoop(); // Start fetching
                } else if (json.type === 'stream' && json.data?.uri) {
                    // Reference implementation prefers messageServer but allows fallback or parallel?
                    // Actually reference `connectMessageWs` waits for `messageServer`.
                    // But if we only get `stream` (like in earlier tests), we might need to use it.
                    // The user said "Clone logic", reference logic uses `messageServer`.
                    // BUT if my previous success was with fallbacks, I should keep fallback logic 
                    // OR maybe headers/cookies are the reason I didn't get messageServer?
                    if (!this.viewUri && (json.data.uri.includes('mpn') || json.data.uri.includes('livedelivery'))) {
                        console.log('[NicoClient] Fallback to Stream URI');
                        this.viewUri = json.data.uri;
                        this.startEntryLoop();
                    }
                }
            } catch (e) {
                console.error(e);
            }
        };

        this.ws.onclose = () => {
            console.log('[NicoClient] WS Close');
            this.active = false;
        };
    }

    // Logic ported from `createEntryFetcher` loop
    private async startEntryLoop() {
        if (!this.viewUri || !this.ndgrRoot) return;

        // Reset state
        this.currentAt = 'now';
        this.abortController = new AbortController();
        const signal = this.abortController.signal;

        console.log(`[NicoClient] Starting Entry Loop from: ${this.currentAt}`);

        try {
            while (this.active) {
                if (signal.aborted) break;

                const separator = this.viewUri.includes('?') ? '&' : '?';
                const uri = `${this.viewUri}${separator}at=${this.currentAt}`;

                console.log(`[NicoClient] Fetching Entry: ${uri}`);

                let receivedSegment = false;

                // Fetch Entry Stream
                const response = await fetch(uri, { signal });
                if (!response.body) throw new Error('No body');

                const reader = response.body.getReader();
                const ChunkedEntry = this.ndgrRoot.lookupType("dwango.nicolive.chat.service.edge.ChunkedEntry");

                await this.readStream(reader, ChunkedEntry, (entry: any) => {
                    // Logic from Reference
                    if (entry.next) {
                        // Case: Next (Update polling cursor)
                        this.currentAt = entry.next.at;
                        // console.log(`[NicoClient] Next At: ${this.currentAt}`);
                    }
                    else if (entry.segment) {
                        // Case: Segment (Real-time data)
                        receivedSegment = true;
                        // console.log(`[NicoClient] Segment URI: ${entry.segment.uri}`);
                        this.fetchMessages(entry.segment.uri);
                    }
                    else if (entry.backward) {
                        // Case: Backward (Past comments)
                        // Reference fetches backward messages here.
                        // For simplicity, we can fetch them or ignore if we only want live.
                        // User said "get comments", usually implies past too if just joined.
                        // But backward logic is complex (recursive). 
                        // For now, let's just log it.
                        console.log('[NicoClient] Backward segment available (Skipping for lighter implementation)');
                    }
                    else if (entry.previous) {
                        // Case: Previous (Recent past?)
                        if (!receivedSegment) {
                            // console.log(`[NicoClient] Previous URI: ${entry.previous.uri}`);
                            this.fetchMessages(entry.previous.uri);
                        }
                    }
                });

                // Loop continues with new `currentAt`
                // If fetching fails or stream closes, we loop again.
                // Ref adds a small sleep/backoff? 
                // Reference `while(true)` -> `fetchEntry` -> calls `ResponseIteratorSet` 
                // It seems to loop immediately.
            }
        } catch (e) {
            console.error('[NicoClient] Entry Loop Terminated', e);
            if (this.active) {
                // Retry with cooling
                setTimeout(() => this.startEntryLoop(), 2000);
            }
        }
    }

    private async fetchMessages(uri: string) {
        if (!this.ndgrRoot) return;
        // console.log(`[NicoClient] Fetching Messages: ${uri}`);

        try {
            const response = await fetch(uri);
            if (!response.body) throw new Error('No body');
            const reader = response.body.getReader();

            const ChunkedMessage = this.ndgrRoot.lookupType("dwango.nicolive.chat.service.edge.ChunkedMessage");

            await this.readStream(reader, ChunkedMessage, (msg: any) => {
                this.handleChunkedMessage(msg);
            });
        } catch (e) {
            console.error('[NicoClient] Message Fetch Error', e);
        }
    }

    private handleChunkedMessage(msg: any) {
        // proto definition:
        // message.payload (oneof) -> message (NicoliveMessage) -> data (oneof) -> chat (Chat)

        // Log to debug structure if needed
        // console.log('ChunkedMsg:', JSON.stringify(msg));

        // In bundled proto, fields are names, not ids usually (if loaded via json) or ids if logic uses decode
        // Our readStream uses `toObject` with options.

        const payload = msg.payload;
        if (!payload) return;

        // Handle Chat
        if (msg.message && msg.message.chat) { // If simplified
            // proto matches: oneof payload { NicoliveMessage message = 2; }
            // NicoliveMessage { oneof data { Chat chat = 1; } }
            // But Wait! `oneof` fields in protobufjs `toObject` are flattened or nested?
            // Usually `payload` key exists with value of oneof option?
            // Or `message` key exists directly?
            // Let's check logic:
            // msg.message corresponds to `NicoliveMessage`.
        }

        // Let's check dynamic access based on proto bundle structure
        let chat = null;

        if (msg.message && msg.message.chat) {
            chat = msg.message.chat;
        } else if (payload === 'message' && msg.message && msg.message.chat) {
            chat = msg.message.chat;
        } else if (msg.payload && msg.payload.message && msg.payload.message.chat) {
            // Access path: ChunkedMessage -> payload(value) -> message -> chat
            // Depending on how `toObject` output.
            // Usually: { message: { chat: { ... } } } if payload='message'
            chat = msg.payload.message.chat;
        } else if (msg.message && msg.message.data && msg.message.data.chat) {
            // If nested data
            chat = msg.message.data.chat;
        }

        // Just defensive check all paths
        const realChat = chat || msg?.message?.chat || msg?.payload?.message?.chat || msg?.payload?.message?.data?.chat;

        if (realChat && realChat.content) {
            const name = realChat.name || 'Anonymous';
            this.onMessage({ author: name, message: realChat.content });
        }
    }

    private async readStream(reader: ReadableStreamDefaultReader<Uint8Array>, msgType: protobuf.Type, callback: (obj: any) => void) {
        let buffer = new Uint8Array(0);

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            const newBuffer = new Uint8Array(buffer.length + value.length);
            newBuffer.set(buffer);
            newBuffer.set(value, buffer.length);
            buffer = newBuffer;

            let offset = 0;
            while (offset < buffer.length) {
                let length = 0;
                let shift = 0;
                let i = offset;
                let hasLength = false;
                while (i < buffer.length) {
                    const byte = buffer[i];
                    length |= (byte & 0x7F) << shift;
                    shift += 7;
                    i++;
                    if ((byte & 0x80) === 0) {
                        hasLength = true;
                        break;
                    }
                }

                if (!hasLength) { break; }

                if (buffer.length - i >= length) {
                    const chunk = buffer.slice(i, i + length);
                    offset = i + length;
                    try {
                        const decoded = msgType.decode(chunk);
                        const obj = msgType.toObject(decoded, { longs: String, enums: String, bytes: String, defaults: true, oneofs: true });
                        callback(obj);
                    } catch (e) { }
                } else { break; }
            }
            if (offset > 0) {
                buffer = buffer.slice(offset);
            }
        }
    }

    leave() {
        this.active = false;
        if (this.ws) {
            this.ws.close();
        }
        if (this.abortController) {
            this.abortController.abort();
        }
    }
}

File: src-tauri/src/niconico.rs
use futures_util::{SinkExt, StreamExt};
use prost::Message;
use regex::Regex;
use reqwest::header::USER_AGENT;
use serde::Serialize;
use serde_json::Value;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::protocol::Message as WsMessage;

// Include generated proto
pub mod ndgr {
    include!(concat!(env!("OUT_DIR"), "/dwango.nicolive.chat.service.edge.rs"));
}

#[derive(Clone, Serialize)]
struct CommentEvent {
    author: String,
    message: String,
}

#[derive(Clone, Serialize)]
struct SystemEvent {
    author: String,
    message: String,
}

pub struct NiconicoClient {
    running: Arc<AtomicBool>,
    app_handle: AppHandle,
}

impl NiconicoClient {
    pub fn new(app_handle: AppHandle) -> Self {
        Self {
            running: Arc::new(AtomicBool::new(false)),
            app_handle,
        }
    }

    pub fn stop(&self) {
        self.running.store(false, Ordering::Relaxed);
    }

    pub fn start(&self, url: String) {
        if self.running.load(Ordering::Relaxed) {
            return;
        }
        self.running.store(true, Ordering::Relaxed);
        let running_flag = self.running.clone();
        let app = self.app_handle.clone();

        tokio::spawn(async move {
            if let Err(e) = connect_and_run(url, app.clone(), running_flag.clone()).await {
                // emit error
                let _ = app.emit("comment", SystemEvent {
                    author: "System".into(),
                    message: format!("Error: {}", e),
                });
            }
            running_flag.store(false, Ordering::Relaxed);
        });
    }
}

use std::collections::HashSet;
use std::sync::Mutex;

async fn connect_and_run(url: String, app: AppHandle, running: Arc<AtomicBool>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let client = reqwest::Client::new();
    
    // 1. Fetch Page
    let res = client.get(&url)
        .header(USER_AGENT, "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
        .send().await?.text().await?;

    // 2. Extract Props
    let re_script = Regex::new(r#"<script[^>]+id="embedded-data"[^>]*>([\s\S]*?)</script>"#).unwrap();
    let re_attr = Regex::new(r#"data-props="([^"]+)""#).unwrap();

    let props_str = if let Some(caps) = re_script.captures(&res) {
        let content = caps[1].trim().to_string();
        if !content.is_empty() {
             app.emit("comment", SystemEvent { author: "System".into(), message: "Found embedded-data in script tag".into() })?;
             content
        } else {
             // Fallback to attribute if script body is empty
             if let Some(caps_attr) = re_attr.captures(&res) {
                 app.emit("comment", SystemEvent { author: "System".into(), message: "Found embedded-data in data-props attribute (fallback)".into() })?;
                 urlencoding::decode(&caps_attr[1])?.replace("&quot;", "\"")
             } else {
                 return Err("Found script tag but no content and no data-props".into());
             }
        }
    } else if let Some(caps) = re_attr.captures(&res) {
        app.emit("comment", SystemEvent { author: "System".into(), message: "Found embedded-data in data-props attribute".into() })?;
        urlencoding::decode(&caps[1])?.replace("&quot;", "\"")
    } else {
        return Err("Could not find embedded-data".into());
    };
    

    let props: Value = match serde_json::from_str(&props_str) {
        Ok(v) => v,
        Err(e) => {
            let snippet = if props_str.len() > 200 { &props_str[..200] } else { &props_str };
            app.emit("comment", SystemEvent { author: "Fatal".into(), message: format!("JSON Parse Error: {} | Content: {}", e, snippet) })?;
            return Err(Box::new(e));
        }
    };
    let ws_url = props["site"]["relive"]["webSocketUrl"].as_str().ok_or("No WebSocket URL")?;
    
    app.emit("comment", SystemEvent { author: "System".into(), message: "WebSocket URL found".into() })?;

    // 3. Connect WebSocket
    let (ws_stream, _) = connect_async(ws_url).await?;
    let (mut write, mut read) = ws_stream.split();

    // Send startWatching
    let start_watching = serde_json::json!({
        "type": "startWatching",
        "data": {
            "stream": {
                "quality": "super_high",
                "latency": "low",
                "chasePlay": false
            },
            "reconnect": false
        }
    });

    write.send(WsMessage::Text(start_watching.to_string())).await?;

    let mut view_uri = String::new();

    // Listen for MessageServer
    while let Some(msg) = read.next().await {
        if !running.load(Ordering::Relaxed) { return Ok(()); }
        let msg = msg?;
        if let WsMessage::Text(text) = msg {
            let json: Value = serde_json::from_str(&text)?;
            if json["type"] == "ping" {
                write.send(WsMessage::Text(serde_json::json!({"type": "pong"}).to_string())).await?;
            } else if json["type"] == "messageServer" {
                view_uri = json["data"]["viewUri"].as_str().unwrap_or("").to_string();
                break;
            } else if json["type"] == "stream" {
                 // Fallback if messageServer not received immediately, but usually messageServer comes first or soon.
                 // If we get stream uri, we can use it as fallback?
                 if let Some(uri) = json["data"]["uri"].as_str() {
                     if view_uri.is_empty() {
                         view_uri = uri.to_string();
                     }
                 }
            }
        }
    }

    if view_uri.is_empty() {
         // Maybe continue reading loop?
         // For now, assume failure if loop broke without URI
         // Re-implement read loop closer to real logic?
         return Err("No View URI received".into());
    }

    app.emit("comment", SystemEvent { author: "System".into(), message: "Connected to Message Server".into() })?;

    // 4. Entry Loop
    let mut current_at = "now".to_string();
    // Cache for deduplication
    let seen_ids = Arc::new(Mutex::new(HashSet::new()));

    while running.load(Ordering::Relaxed) {
        let separator = if view_uri.contains("?") { "&" } else { "?" };
        // Better:
        let uri = format!("{}{}at={}", view_uri, separator, current_at);
        
        let mut resp = client.get(&uri).send().await?;
        
        // Stream body
        let mut buffer = temp_niconama::Buffer::new();
        
        while let Some(chunk) = resp.chunk().await? {
            if !running.load(Ordering::Relaxed) { break; }
            buffer.extend_from_slice(&chunk);
            
            // Decode loop
            while let Some(entry) = buffer.decode_next::<ndgr::ChunkedEntry>() {
                if let Some(entry_data) = entry.entry {
                    match entry_data {
                        ndgr::chunked_entry::Entry::Next(next) => {
                           current_at = next.at.to_string();
                        },
                        ndgr::chunked_entry::Entry::Segment(seg) => {
                             let seg_uri = seg.uri;
                             let app_clone = app.clone();
                             let client_clone = client.clone();
                             let seen_ids_clone = seen_ids.clone();
                             tokio::spawn(async move {
                                 if let Err(e) = fetch_segment(client_clone, seg_uri, app_clone.clone(), seen_ids_clone).await {
                                     // Log error inside spawn
                                     let _ = app_clone.emit("comment", SystemEvent { author: "Debug".into(), message: format!("Segment Error: {}", e) }); 
                                 }
                             });
                        },
                        ndgr::chunked_entry::Entry::Previous(_prev) => {
                             // Ignore previous segments to reduce latency and focus on live comments for now.
                             // let seg_uri = prev.uri;
                             // ... spawn ...
                        },
                         _ => {}
                    }
                }
            }
        }
    }

    Ok(())
}

async fn fetch_segment(client: reqwest::Client, uri: String, app: AppHandle, seen_ids: Arc<Mutex<HashSet<String>>>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut resp = client.get(&uri).send().await?;
    
    // Stream processing instead of wait-for-all
    let mut buffer = temp_niconama::Buffer::new();
    
    while let Some(chunk) = resp.chunk().await? {
        buffer.extend_from_slice(&chunk);
        
        while let Some(msg) = buffer.decode_next::<ndgr::ChunkedMessage>() {
             // Deduplication
             if let Some(meta) = &msg.meta {
                 let mut ids = seen_ids.lock().unwrap();
                 if ids.contains(&meta.id) {
                     continue;
                 }
                 ids.insert(meta.id.clone());
             }
             
             if let Some(payload) = msg.payload {
                 match payload {
                     ndgr::chunked_message::Payload::Message(nico_msg) => {
                         if let Some(data) = nico_msg.data {
                             match data {
                                 ndgr::nicolive_message::Data::Chat(chat) => {
                                     let name = if !chat.name.is_empty() {
                                         chat.name
                                     } else if !chat.hashed_user_id.is_empty() {
                                         chat.hashed_user_id
                                     } else {
                                         "Anonymous".to_string()
                                     };
                                     let _ = app.emit("comment", CommentEvent {
                                         author: name,
                                         message: chat.content,
                                     });
                                 },
                                 _ => {}
                             }
                         }
                     },
                     _ => {}
                 }
             }
        }
    }
    
    Ok(())
}

// Helper module for buffer and varint decoding
mod temp_niconama {
    use prost::Message;
    
    pub struct Buffer {
        data: Vec<u8>,
    }
    
    impl Buffer {
        pub fn new() -> Self {
            Self { data: Vec::new() }
        }
        
        pub fn extend_from_slice(&mut self, other: &[u8]) {
            self.data.extend_from_slice(other);
        }
        
        pub fn decode_next<T: Message + Default>(&mut self) -> Option<T> {
             // Try to read a length prefix (varint)
             // Simple varint reader implementation
             let mut i = 0;
             let mut len: u64 = 0;
             let mut shift = 0;
             
             loop {
                 if i >= self.data.len() { return None; } // Not enough bytes
                 let b = self.data[i];
                 len |= ((b & 0x7F) as u64) << shift;
                 shift += 7;
                 i += 1;
                 if b & 0x80 == 0 { break; }
             }
             
             let msg_len = len as usize;
             if self.data.len() < i + msg_len {
                 return None; // Waiting for more data
             }
             
             // Decode
             let msg_buf = &self.data[i..i+msg_len];
             let msg = T::decode(msg_buf).ok();
             
             // Advance buffer
             self.data.drain(0..i+msg_len);
             
             msg
        }
    }
}
